### PR TITLE
[specs][kimchi] index and constraint system parts

### PR DESCRIPTION
**Pull request by mimoo**
To make it easier to spec the constraint system and index setup:

* index -> prover_index and verifier_index
* index::linearization -> linearization

This also adds some spec comment in order to fill the spec out. More to come.

### CODE REVIEW

**Review by jspada — APPROVED** [review]
just moving things around.  LGTM!

File: book/specifications/kimchi/template.md [hunk 1]
```diff
@@ -88,9 +88,17 @@ TODO: use expr to define the index columns?
...
+## Prover Index
 
-{sections.indexes}
+{sections.prover_index}
+
+## Verifier Index
+
+The verifier index is essentially a number of pre-computations containing:
+
+* the (non-hidding) commitments of all the required polynomials
```
Comment (querolita) [inline book/specifications/kimchi/template.md:99]: are these polynomials, partly, each of the gate constraints of the specific circuit? Meaning, the specification of the circuit, which is public information (thus non-hiding commitments being used).
Typo -> non-hiding, only 1 d I think

File: book/src/specs/kimchi.md [hunk 2]
```diff
@@ -188,9 +188,28 @@ fifth round:
 
 ## Constraint System Creation
 
+1. +3 on gates.len() here to ensure that we have room for the zero-knowledge entries of the permutation polynomial
+   see https://minaprotocol.com/blog/a-more-efficient-approach-to-zero-knowledge-for-plonk
+2. pad the rows: add zero gates to reach the domain size
 
 
-## Prover and Verifier Index Creation
+## Prover Index
```
Comment (querolita) [inline book/src/specs/kimchi.md:196]: If I understand correctly, this index is dependent on the circuit itself? Like, this prover index is made effective only after the "general" SRS has been created? ( I refer to this question in the other PR as well, making a distinction between Setup and Index in papers)

File: book/src/specs/kimchi.md [hunk 3]
```diff
@@ -188,9 +188,28 @@ fifth round:
...
+   which is obtained by looking at the highest monomial in the sum
+    $$\sum_{i=0}^{PERMUTS} (w_i(x) + \beta k_i x + \gamma)$$
+   where the $w_i(x)$ are of degree the size of the domain.
+
+
+## Verifier Index
+
+The verifier index is essentially a number of pre-computations containing:
+
+* the (non-hidding) commitments of all the required polynomials
```
Comment (querolita) [inline book/src/specs/kimchi.md:212]: non hiding

File: kimchi/src/linearization.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
+//! This module implements the linearization.
```
Comment (querolita) [inline kimchi/src/linearization.rs:1]: I like this to have a separate file

**Review by querolita — APPROVED** [review]
I like this modular approach

File: book/src/specs/kimchi.md (hunk 2 above)
Comment (mimoo) [reply book/src/specs/kimchi.md:196]: > If I understand correctly, this index is dependent on the circuit itself?

Yeah you basically compile the circuit to these two indexes

File: book/specifications/kimchi/template.md (hunk 1 above)
Comment (querolita) [reply book/specifications/kimchi/template.md:99]: Oh! so only those polynomials that will tell the verifier what gate appears in each row I see... The mutually exclusive polynomials you mean.

Well I have one question. I understand that the verifier checks that `quot(X) * van(X) = poly(x)`, where `poly` is the set of all constraints, permutations etc. Then, how does it check that `poly(X)` indeed corresponds to "the shape that it should"? Will it "rebuild" the equation checks using the committed witness together with the set of constraints? (but perhaps this approach takes linear time in the number of rows, right?) I don't know if I am making myself clear.

Comment (mimoo) [reply book/specifications/kimchi/template.md:99]: that's what I call the polynomial dance between the prover and the verifier. The verifier doesn't want to just get poly (or f) from the prover, the verifier recreates an f from the constraints of each gate, the witness, the permutation polynomial, etc. Oh, and they do that hidden in the exponent, so as a commitment, not as a polynomial directly.

**Pull request merged**